### ISSUE TITLE
How to retrieve a String data as a return value

### ISSUE THREAD

**Issue by cybafelo** [collectd/collectd#2764]
Please make it possible to read string values from SNMP plugin

**Comment by faxm0dem** [comment]
How would you want  to use the returned string?
Collectd collects numbers, and there is no such thing as a string data type.

**Comment by cybafelo** [comment]
yes i understand - its based on the types.db which exclusively deals with numbers - however SNMP returns not only numbers but also strings - and what happens in this case - collectd snmp plugin successfully reads the string data for example an ip address in the form "192.168.3.34", or a MAC address like "A4:35:23:45:55:22" or DHCP hostname which is "nasty-computer-on-my-network"  - which is clearly very useful information
 
but then it tries to convert it to a number - (GAUGE in my case since i don't want to modify the data)

At this point - snmp.c throws an error at line 898

`status = parse_value(string, &ret, type);`
      `if (status != 0) {`
       `ERROR("snmp plugin: host %s: csnmp_value_list_to_value: Parsing string "`
              `"as %s failed: %s",`
             `(host_name != NULL) ? host_name : "UNKNOWN",`
              `DS_TYPE_TO_STRING(type), string);`
      `}`

and then later converts it to NAN

`case DS_TYPE_GAUGE:`
        `ret.gauge = NAN;`
        `break;`

an my argument is - we are now in 2018 - surely we can allocate some memory - introduce a new types.db value, call it STRING, allocate some memory - set the return value accordingly and free it afterwards?

I quote Robert Cowart:

_"I probably prefer Telegraf to Collectd, but I don't think either are great as neither does the things necessary for truly efficient SNMP collection. In my opinion a truly scalable and consumable SNMP poller is one of the things missing in the world of open source. They are all compromised in one way or the other. This hole in open source is probably the only reason that a lot of legacy NMS vendors are still alive."_

and in my opinion - its because collectd is stuck in 1980 where numbers were only important - this addition of a 'string' type (overwrite it ffs in rrd if you have to) will hurt no one - and if you don't want to introduce a backward compatible change - then at least store the value as a gauge and clean it up afterwards.

**Comment by cybafelo** [comment]
i mean i could attempt to do it too but it i'm sure someone with an expert opinion could look at this and do it quite easily.. ?

**Comment by faxm0dem** [comment]
What you are suggesting (not very nicely/politely/calmly) would require a lot of design changes.
That being said, collectd is moving (slowly but calmly) towards implementing arbitrary key/value attrributes attached to each value. This would free the user from the (admittedly a bit 90's style) static and limited host/plugin/instance/type paradigm.

In this particular use-case you're referring to, it would mean you could collect any snmp string data and add it as key/value pairs to any collected number. For instance you could have the interface throughput with the physical address and ifAlias for instance.

This change has already started, it's a lot of work in the code base, but it won't allow the values themselves to be strings. That will probably never happen, by design (correct me @octo if I'm wrong). How would you for instance handle thresholds for strings?

I quote Abraham Maslow:

"If you only have a hammer you tend to see every problem as a screw"

**Comment by rpv-tomsk** [comment]
Once I thought about such feature for Collectd.
That may be very useful in monitoring systems to use string data, received from different sources.
Such data usually changes very rarely and can't be represented as numeric metrics.

Exact string value needs to be transferred from data source to outputs/storages.
Because of this, such data can be processed only by a completely separate subsystem of Collectd, if that will be eventually implemented.
Such string values may not be associated with metrics and may be processed as completely independent.

Many existing plugins can act as a source for string values.
Examples of string values that could be widely used:

- vendor, model, serial number of devices (`cpu`, `smart`, `ipmi` and others 'hardware' plugins).
- uptime (`uptime` plugin)
- RAID state, RAID level (`md` plugin)
- UPS state (`apcups`, `nut` plugins)
- connection states (different plugins, which monitors services which Collectd can lose connection to: `apache`, `memcached` and so on)

Such string values can also be formed based on existing metrics, using thresholds.
Notifications can be send when string value changes.

As the `Collect-D` name implies, Collectd will collect and deliver these string values - this is the Collectd was intended for, I think.

That was the thoughts I have.

**Comment by jacobisaliveandwell** [comment]
2020. any updates here?

**Comment by faxm0dem** [comment]
What are you trying to achieve exactly @jacobisaliveandwell ?

**Comment by jacobisaliveandwell** [comment]
@faxm0dem I've solved this with an exec, but it would be pretty nice to return 1/0 for regex matches within json.

**Comment by faxm0dem** [comment]
I think this would be too narrow a use-case.
This is exactly why collectd has exec, python, perl, golang integrations so that people can easily meet their needs

**Comment by jacobisaliveandwell** [comment]
I think generalized support for substring replacement could have many applications.

Farming out to exec can lead to spaghetti management. Don't get me wrong, it's definitely the right answer for a number of cases, but it adds to maintenance exponentially.

The maintainers are welcome to close this issue.

**Comment by octo** [comment]
This sounds similar to the "tail plugin" and could possibly trust the "libmatch.la" library used by that plugin.
